content/blog/2022/2022-04-26-introduction-to-retrofit.md

```diff
@@ -0,0 +1,536 @@
+
+    @JsonProperty("id")
+    private long id;
+
+    @JsonProperty("name")
+    private String name;
+
+    @JsonProperty("dob")
+    private String dob;
+
+}
+````
+
+````java
+@Getter
+@Setter
+@NoArgsConstructor
+public class BookDto {
+    @JsonProperty("bookId")
+    private long id;
+
+    @JsonProperty("bookName")
+    private String name;
+
+    @JsonProperty("publisher")
+    private String publisher;
+
+    @JsonProperty("publicationYear")
+    private String publicationYear;
+
+    @JsonProperty("isCopyrighted")
+    private boolean copyrightIssued;
+
+    @JsonProperty("authors")
+    private Set<AuthorDto> authors;
+}
+
+````
+````java
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class LibResponse {
+    private String responseCode;
+
+    private String responseMsg;
+}
+````
+
+### Creating the client interface
+To create the retrofit interface, we will map every service call with a corresponding interface method as shown in the screenshot below.
+{{% image alt="settings" src="images/posts/retrofit/REST.jpg" %}}
+
+````java
+public interface LibraryClient {
+
+    @GET("/library/managed/books")
+    Call<List<BookDto>> getAllBooks(@Query("type") String type);
+
+    @POST("/library/managed/books")
+    Call<LibResponse> createNewBook(@Body BookDto book);
+
+    @PUT("/library/managed/books/{id}")
+    Call<LibResponse> updateBook(@Path("id") Long id, @Body BookDto book);
+
+    @DELETE("/library/managed/books/{id}")
+    Call<LibResponse> deleteBook(@Path("id") Long id);
+}
+````
+We will deep-dive into the annotations and Retrofit classes in the further sections.
+
+### Creating a Retrofit.Builder class as a Spring Boot Configuration 
+We will use the Retrofit Builder API to define URL for HTTP operations.
+
+````java
+@Configuration
+@EnableConfigurationProperties(ClientConfigProperties.class)
+public class RestClientConfiguration {
+
+    @Bean
+    public LibraryClient libraryClient(ClientConfigProperties props) {
+        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder()
+                .addInterceptor(new BasicAuthInterceptor(props.getUsername(), props.getPassword()))
+                .connectTimeout(props.getConnectionTimeout(), TimeUnit.SECONDS)
+                .readTimeout(props.getReadWriteTimeout(), TimeUnit.SECONDS);
+
+        return new Retrofit.Builder().client(httpClientBuilder.build())
+                .baseUrl(props.getEndpoint())
+                .addConverterFactory(JacksonConverterFactory.create(new ObjectMapper()))
+                .build().create(LibraryClient.class);
+
+    }
+
+}
+````
+
+## Using Retrofit
+In the further sections we will learn more about the Retrofit API and how to use them. 
+
+### Building a Client Interface
+In this section, we will look at how to build the client interface.
+**Retrofit supports annotations @GET, @POST, @PUT, @DELETE, @PATCH, @OPTIONS, @HEAD** which we use to annotate our client methods as shown below
+
+````java
```

Comment:
Please make each group of the examples its own section (except the first, maybe), even if that section hast only one paragraph. I.e.

* Path parameters
* Query parameters
* Multiple Query Parameters
* Request body
* Headers
*